text/3626-format-args-implicit-dot.md

```diff
@@ -0,0 +1,202 @@
+
+# Motivation
+[motivation]: #motivation
+
+[RFC 2795](https://github.com/rust-lang/rfcs/pull/2795) added "implicit named
+arguments" to `std::format_args!` (and other macros based on it such as
+`format!` and `println!` and `panic!`), allowing the format string to reference
+variables in scope using identifiers. For instance, `println!("Hello {name}")`
+is now equivalent to `println!("Hello {name}", name=name)`.
+
+The original implicit named arguments mechanism only permitted single
+identifiers, to avoid the complexity of embedding arbitrary expressions into
+format strings. The implicit named arguments mechanism is widely used, and one
+of the most common requests and most common reasons people cannot use that
+syntax is when they need to access a struct field. Adding struct field syntax
+does not conflict with any other format syntax, and unlike allowing *arbitrary*
+expressions, allowing struct field syntax does not substantially increase
+complexity or decrease readability.
+
+This proposal has the same advantages as the original implicit named arguments
+proposal: making more formatting expressions easy to read from left-to-right
+without having to jump back and forth between the format string and the
+arguments.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+With this proposal accepted, the following (currently invalid) macro
+invocation:
+
+```rust
+format_args!("hello {person.name}")
+```
+
+would become a valid macro invocation, and would be equivalent to a shorthand
+for the already valid:
+
+```rust
+format_args!("hello {unique_ident}", unique_ident=person.name)
+```
+
+The identifier at the beginning of the chain (`person` in this case) must be an
+identifier which existed in the scope in which the macro is invoked, and must
+have a field of the appropriate name (`name` in this case).
+
+This syntax works for fields within fields as well:
+
+```rust
+format_args!("{obj.field.nested_field.another_field}")
+```
+
+As a result of this change, downstream macros based on `format_args!` would
+also be able to accept implicit named arguments in the same way. This would
+provide ergonomic benefit to many macros across the ecosystem, including:
+
+ - `format!`
+ - `print!` and `println!`
+ - `eprint!` and `eprintln!`
+ - `write!` and `writeln!`
+ - `panic!`, `unreachable!`, `unimplemented!`, and `todo!`
+ - `assert!`, `assert_eq!`, and similar
+ - macros in the `log` and `tracing` crates
+
+(This is not an exhaustive list of the many macros this would affect.)
+
+## Additional formatting parameters
+
+As a result of this RFC, formatting parameters can also use implicit named
+argument capture:
+
+```rust
+println!("{self.value:self.width$.self.precision$}");
+```
+
+This is slightly complex to read, but unambiguous thanks to the `$`s.
+
+## `await`
+
+Formatting can use `.await`, as well:
+
+```rust
+println!("{future1.await} {future2.await}");
```

Comment:
since `.await` is definitely having side effect (technically `obj.field` can also have side effect with some malicious Deref impl) perhaps this RFC should specify the order and number of times an expression will be evaluated.

```rust
println!("{a.await}, {a.await}, {b.await}, {c.await:d.await$.e.await$}, {hh}, {f.await}", hh=g.await);
```

Intuitively everything should be evaluated left-to-right and as many time as specified

```rust
println!("{x0}, {x1}, {x2}, {x3:x4$.x5$}, {hh}, {x6}",
  x0=a.await,
  x1=a.await,
  x2=b.await,
  x3=c.await,
  x4=d.await,
  x5=e.await,
  x6=f.await,
  hh=g.await,
);
```

Reply:
@kennytm There's already a mention of evaluation order in the reference section:

> Any `Deref` operations or `.await` operations associated with the `.` in each
format argument are evaluated from left-to-right as they appear in the format
string, at the point where the format string argument is evaluated, before the
positional or named arguments are evaluated.

I'll add a mention of evaluating exactly once.